### PR TITLE
Change some of the handling and add to it some extra packets needed!

### PR DESCRIPTION
**Pull request by VennDev**
++ ChangeDimensionPacket
++ MobArmorEquipmentPacket
++ PlayerArmorDamagePacket

### CODE REVIEW

File: src/Supero/NightfallProtocol/network/packets/ChangeDimensionPacket.php [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+{
+    public int $dimension;
+    public Vector3 $position;
+    public bool $respawn = false;
+    private ?int $loadingScreenId = null;
+
+    /**
+     * @generate-create-func
+     */
+    public static function create(int $dimension, Vector3 $position, bool $respawn, ?int $loadingScreenId): self{
```
Comment (Superomarking) [inline src/Supero/NightfallProtocol/network/packets/ChangeDimensionPacket.php]: Rename to `createPacket`

File: src/Supero/NightfallProtocol/network/packets/ChangeDimensionPacket.php [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        $result->respawn = $respawn;
+        $result->loadingScreenId = $loadingScreenId;
+        return $result;
+    }
+
+    protected function decodePayload(PacketSerializer $in): void {
+        $this->dimension = $in->getVarInt();
+        $this->position = $in->getVector3();
+        $this->respawn = $in->getBool();
+        if($in->getProtocolId() >= CustomProtocolInfo::PROTOCOL_1_21_20){
```
Comment (Superomarking) [inline src/Supero/NightfallProtocol/network/packets/ChangeDimensionPacket.php]: It's `getProtocol()` not `getProtocolId`

File: src/Supero/NightfallProtocol/network/packets/ChangeDimensionPacket.php [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+            $this->loadingScreenId = $in->readOptional(fn() => $in->getLInt());
+        }
+    }
+
+    protected function encodePayload(PacketSerializer $out): void {
+        $out->putVarInt($this->dimension);
+        $out->putVector3($this->position);
+        $out->putBool($this->respawn);
+        if($out->getProtocolId() >= CustomProtocolInfo::PROTOCOL_1_21_20){
```
Comment (Superomarking) [inline src/Supero/NightfallProtocol/network/packets/ChangeDimensionPacket.php]: Same over here

File: src/Supero/NightfallProtocol/network/packets/ChangeDimensionPacket.php [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+            $out->writeOptional($this->loadingScreenId, $out->putLInt(...));
+        }
+    }
+
+    public function handle(PacketHandlerInterface $handler): bool {
```
Comment (Superomarking) [inline src/Supero/NightfallProtocol/network/packets/ChangeDimensionPacket.php]: This isn't needed

File: src/Supero/NightfallProtocol/network/packets/PlayerArmorDamagePacket.php [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    private ?int $headSlotDamage;
+    private ?int $chestSlotDamage;
+    private ?int $legsSlotDamage;
+    private ?int $feetSlotDamage;
+    private ?int $bodySlotDamage;
+
+    /**
+     * @generate-create-func
+     */
+    public static function create(?int $headSlotDamage, ?int $chestSlotDamage, ?int $legsSlotDamage, ?int $feetSlotDamage, ?int $bodySlotDamage) : self{
```
Comment (Superomarking) [inline src/Supero/NightfallProtocol/network/packets/PlayerArmorDamagePacket.php]: Same over here

File: src/Supero/NightfallProtocol/network/packets/PlayerArmorDamagePacket.php [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    protected function decodePayload(PacketSerializer $in) : void{
+        $flags = $in->getByte();
+
+        $this->headSlotDamage = $this->maybeReadDamage($flags, self::FLAG_HEAD, $in);
+        $this->chestSlotDamage = $this->maybeReadDamage($flags, self::FLAG_CHEST, $in);
+        $this->legsSlotDamage = $this->maybeReadDamage($flags, self::FLAG_LEGS, $in);
+        $this->feetSlotDamage = $this->maybeReadDamage($flags, self::FLAG_FEET, $in);
+        if($in->getProtocolId() >= CustomProtocolInfo::PROTOCOL_1_21_20){
```
Comment (Superomarking) [inline src/Supero/NightfallProtocol/network/packets/PlayerArmorDamagePacket.php]: Same over here

File: src/Supero/NightfallProtocol/network/packets/PlayerArmorDamagePacket.php [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+            $this->composeFlag($this->legsSlotDamage, self::FLAG_LEGS) |
+            $this->composeFlag($this->feetSlotDamage, self::FLAG_FEET) |
+            ($out->getProtocolId() >= CustomProtocolInfo::PROTOCOL_1_21_20 ? $this->composeFlag($this->bodySlotDamage, self::FLAG_BODY) : 0)
+        );
+
+        $this->maybeWriteDamage($this->headSlotDamage, $out);
+        $this->maybeWriteDamage($this->chestSlotDamage, $out);
+        $this->maybeWriteDamage($this->legsSlotDamage, $out);
+        $this->maybeWriteDamage($this->feetSlotDamage, $out);
+        if($out->getProtocolId() >= CustomProtocolInfo::PROTOCOL_1_21_20){
```
Comment (Superomarking) [inline src/Supero/NightfallProtocol/network/packets/PlayerArmorDamagePacket.php]: and here

File: src/Supero/NightfallProtocol/network/packets/PlayerArmorDamagePacket.php [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+            $this->maybeWriteDamage($this->bodySlotDamage, $out);
+        }
+    }
+
+    public function handle(PacketHandlerInterface $handler) : bool{
```
Comment (Superomarking) [inline src/Supero/NightfallProtocol/network/packets/PlayerArmorDamagePacket.php]: Same thing

**Review by Superomarking — CHANGES_REQUESTED** [review]

**Review by VennDev — COMMENTED** [review]
https://github.com/Nightfall-MCPE/NightfallProtocols/pull/4/commits/d14e964cfb7fbfde14c8baee83f89ef02a5c61f3

File: src/Supero/NightfallProtocol/network/packets/ChangeDimensionPacket.php (hunk 4 above)
Comment (VennDev) [reply src/Supero/NightfallProtocol/network/packets/ChangeDimensionPacket.php]: Has removed!

File: src/Supero/NightfallProtocol/network/packets/ChangeDimensionPacket.php [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+use pocketmine\network\mcpe\protocol\serializer\PacketSerializer;
+use pocketmine\network\mcpe\protocol\ChangeDimensionPacket as PM_Packet;
+use Supero\NightfallProtocol\utils\ReflectionUtils;
+
+class ChangeDimensionPacket extends PM_Packet
+{
+    public int $dimension;
+    public Vector3 $position;
+    public bool $respawn = false;
+    public ?int $_loadingScreenId = null;
```
Comment (Superomarking) [inline src/Supero/NightfallProtocol/network/packets/ChangeDimensionPacket.php]: Whats the need for the underscore?

File: src/Supero/NightfallProtocol/network/packets/MobArmorEquipmentPacket.php [hunk 10]
```diff
@@ -49,14 +50,19 @@ protected function encodePayload(PacketSerializer $out) : void {
...
+     */
     public function getConstructorArguments(PM_Packet $packet): array {
-        return [
-            $packet->actorRuntimeId,
-            $packet->head,
-            $packet->chest,
-            $packet->legs,
-            $packet->feet,
-            $packet->body,
+        $properties = [
```
Comment (Superomarking) [inline src/Supero/NightfallProtocol/network/packets/MobArmorEquipmentPacket.php]: Why use reflection here?

File: src/Supero/NightfallProtocol/network/packets/PlayerArmorDamagePacket.php [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    protected function decodePayload(PacketSerializer $in) : void{
+        $flags = $in->getByte();
+
+        $this->headSlotDamage = $this->maybeReadDamage($flags, self::FLAG_HEAD, $in);
+        $this->chestSlotDamage = $this->maybeReadDamage($flags, self::FLAG_CHEST, $in);
+        $this->legsSlotDamage = $this->maybeReadDamage($flags, self::FLAG_LEGS, $in);
+        $this->feetSlotDamage = $this->maybeReadDamage($flags, self::FLAG_FEET, $in);
+        if($in->getProtocolId() >= CustomProtocolInfo::PROTOCOL_1_21_20){
```
Comment (Superomarking) [inline src/Supero/NightfallProtocol/network/packets/PlayerArmorDamagePacket.php]: getProtocolId() -> getProtocol()

File: src/Supero/NightfallProtocol/network/packets/PlayerArmorDamagePacket.php [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+            $this->maybeWriteDamage($this->bodySlotDamage, $out);
+        }
+    }
+
+    /**
+     * @throws ReflectionException
+     */
+    public function getConstructorArguments(PM_Packet $packet): array
+    {
+        $properties = [
```
Comment (Superomarking) [inline src/Supero/NightfallProtocol/network/packets/PlayerArmorDamagePacket.php]: Again, no need for reflection

**Review by Superomarking — CHANGES_REQUESTED** [review]

File: src/Supero/NightfallProtocol/network/packets/MobArmorEquipmentPacket.php (hunk 10 above)
Comment (VennDev) [reply src/Supero/NightfallProtocol/network/packets/MobArmorEquipmentPacket.php]: oh my IDE wrong

Comment (VennDev) [reply src/Supero/NightfallProtocol/network/packets/MobArmorEquipmentPacket.php]: wait me

File: src/Supero/NightfallProtocol/network/packets/PlayerArmorDamagePacket.php (hunk 12 above)
Comment (VennDev) [reply src/Supero/NightfallProtocol/network/packets/PlayerArmorDamagePacket.php]: image

Comment (VennDev) [reply src/Supero/NightfallProtocol/network/packets/PlayerArmorDamagePacket.php]: It is type private ?

Comment (VennDev) [reply src/Supero/NightfallProtocol/network/packets/PlayerArmorDamagePacket.php]: image

Comment (Superomarking) [reply src/Supero/NightfallProtocol/network/packets/PlayerArmorDamagePacket.php]: use $packet->get(Field)()

Comment (Superomarking) [reply src/Supero/NightfallProtocol/network/packets/PlayerArmorDamagePacket.php]: image

Comment (VennDev) [reply src/Supero/NightfallProtocol/network/packets/PlayerArmorDamagePacket.php]: Ow, sr i dont see it !! :))

Comment (VennDev) [reply src/Supero/NightfallProtocol/network/packets/PlayerArmorDamagePacket.php]: https://github.com/Nightfall-MCPE/NightfallProtocols/pull/4/commits/ded0d1db70e9d0a0b17facc4cd228310d3ac5b45

**Review by Superomarking — APPROVED** [review]

**Pull request merged**